### DISCUSSION TITLE
How to quickly and efficiently modify the values in NDArray object without matrix operations.

### DISCUSSION THREAD

**Discussion by huaizhe2012** [deepjavalibrary/djl#2184] [Q&A] [answered]
I need to perform some calculations to modify the NDArray object according to the conditions through the neural network weight. This requires three layers of loops to continuously modify each element in the NDArray object. It runs very slowly, and the memory usage grows rapidly, and it will overflow the memory after running for a while. In the same case, if I use python, I will get the calculation result very quickly. How can I change the way the NDArray object modifies values? The following is the java code and the corresponding python code.
![python](https://user-images.githubusercontent.com/49636762/203892384-59a04970-ad18-45c3-9e0a-bfe78198cadd.png)
![java](https://user-images.githubusercontent.com/49636762/203892404-852f5b7d-ae02-4820-8880-25c6bcc78791.png)

**Comment by frankfliu** [comment, marked as answer]
@huaizhe2012 

Java doesn't allow operator override, so you won't be able to use index to modify NDArray.

In each operation of NDArray, it actually generates a temporary native NDArray, to avoid OOM, you need to close those temporary resource in each loop manually.

Each set/get call in NDArray has JNI call overhead, it's not efficient. You can do calculation in java heap and then set the hold data to NDArray.

**Reply by huaizhe2012** [reply]
Thanks a lot, I'll give it a try
